### ISSUE TITLE
Unable to use xml to suppress warnings in a method with a modreq parameter

### ISSUE THREAD

**Issue by eerhardt** [mono/linker#2006]
ASP.NET has the [following method](https://github.com/dotnet/aspnetcore/blob/52eff90fbcfca39b7eb58baad597df6a99a542b0/src/Components/WebAssembly/JSInterop/src/WebAssemblyJSRuntime.cs#L60):

```C#
        protected override void EndInvokeDotNet(DotNetInvocationInfo callInfo, in DotNetInvocationResult dispatchResult)
        {
            // For failures, the common case is to call EndInvokeDotNet with the Exception object.
            // For these we'll serialize as something that's useful to receive on the JS side.
            // If the value is not an Exception, we'll just rely on it being directly JSON-serializable.
            var resultOrError = dispatchResult.Success ? dispatchResult.Result : dispatchResult.Exception!.ToString();


            // We pass 0 as the async handle because we don't want the JS-side code to
            // send back any notification (we're just providing a result for an existing async call)
            var args = JsonSerializer.Serialize(new[] { callInfo.CallId, dispatchResult.Success, resultOrError }, JsonSerializerOptions);
            BeginInvokeJS(0, "DotNet.jsCallDispatcher.endInvokeDotNetFromJS", args, JSCallResultType.Default, 0);
        }
```

This results in the following IL:

```
.method family hidebysig virtual 
	instance void EndInvokeDotNet (
		valuetype [Microsoft.JSInterop]Microsoft.JSInterop.Infrastructure.DotNetInvocationInfo callInfo,
		[in] valuetype [Microsoft.JSInterop]Microsoft.JSInterop.Infrastructure.DotNetInvocationResult& modreq([System.Private.CoreLib]System.Runtime.InteropServices.InAttribute) dispatchResult
	) cil managed 
{
```

I'm trying to add an xml warning suppression for this method because it is calling into JsonSerializer (which is `RequiresUnreferencedCode`). However, all combinations of signatures I can try don't appear to work. When I pass `--generate-warning-suppressions xml` to the linker, it spits out the following format:

```xml
<?xml version="1.0" encoding="utf-8"?>
<linker>
  <assembly fullname="Microsoft.JSInterop.WebAssembly, Version=42.42.42.42, Culture=neutral, PublicKeyToken=adb9793829ddae60">
    <attribute fullname="System.Diagnostics.CodeAnalysis.UnconditionalSuppressMessageAttribute">
      <argument>ILLink</argument>
      <argument>IL2026</argument>
      <property name="Scope">member</property>
      <property name="Target">M:Microsoft.JSInterop.WebAssembly.WebAssemblyJSRuntime.EndInvokeDotNet(Microsoft.JSInterop.Infrastructure.DotNetInvocationInfo,Microsoft.JSInterop.Infrastructure.DotNetInvocationResult&amp; modreq(System.Runtime.InteropServices.InAttribute))</property>
    </attribute>
  </assembly>
</linker>
```

But this doesn't work. In other cases that use `in` parameters, I had to append `@` after the type name, but that doesn't work. Neither does dropping the `modreq` part in the string.

Debugging into it, it appears that Cecil is returning the modreq as part of the TypeReference.Name:

![image](https://user-images.githubusercontent.com/8291187/116625341-13fec780-a90f-11eb-8023-9c820b3d261b.png)

And for some reason, that string isn't matching in our parsing logic.

cc @vitek-karas

**Comment by vitek-karas** [comment]
Debugged into this.

How to repro:
* The method must be an override of a virtual
* The parameter type is a struct with the `in` keyword

(Note that without it being virtual this works just fine).
The problem is that the Cecil's representation of this is `RequiredModifierType` for which we don't have a special case in the signature visitor: https://github.com/mono/linker/blob/307f6b0afd2fea639fffa97e67155bafe6aa8461/src/linker/Linker/DocumentationSignatureGenerator.PartVisitor.cs#L132

This means we end up calling Cecil's `Name` property which returns `TypeName& modreq(...)`. Note that there's no namespace and the `modreq()` suffix. 

But the parameter parser stops on the parenthesis after `modreq` and so even if the exact text is specified, it only sees `TypeName& modreq` which doesn't match the expected type name.

**Comment by vitek-karas** [comment]
I actually don't know what the `modreq` is... but if it's not something important we should simply strip it.

**Comment by agocke** [comment]
In theory these modreqs could be used to determine IL verifiability -- I'd keep them on and try to fix this some other way.

**Comment by vitek-karas** [comment]
Sorry - I don't want linker to remove them... but we should not consider them when comparing the Target property of the attribute to the method in code. Assuming they don't affect overload resolution rules (which I doubt).

**Comment by agocke** [comment]
In this case, you're fine, C# won't let you overload `ref` vs `in` parameters, but that's not always true. It may be good enough to strip them and then do a sanity check that there isn't a member with the same name.

**Comment by MichalStrehovsky** [comment]
> Assuming they don't affect overload resolution rules (which I doubt).

They do affect the overload resolution rules. It would matter if we were to do something for a language like C++.

But I think we should ignore them for simplicity sake. It's the same kind of edge case as having two overloads taking types with the same names, but from different assemblies. We can't represent that either.

The bottom line is we should also get out of the business of "oh, just let Cecil come up with a name for this" because Cecil comes up with weird names that don't match anything anyone might have ever seen before (ILASM, C# syntax, reflection notation, ECMA-335 SerString, documentation syntax, etc.). I had my fun with it that I had to vent in comments (and Jan didn't let me [check in](https://github.com/dotnet/runtimelab/pull/422#pullrequestreview-550307479)). We should check how this works for methods taking function pointers.

**Comment by agocke** [comment]
> But I think we should ignore them for simplicity sake. It's the same kind of edge case as having two overloads taking types with the same names, but from different assemblies. We can't represent that either.

I agree, the circumstances where the modreq is the only difference should be very rare (knowing the current way that modreqs are used) and if it becomes more common we can address it later.
